internal/database/postgres.go

```diff
@@ -720,3 +720,148 @@ func (conn PostgresConn) RemoveFile(remove broadcast.RemoveFile) error {
 	}
 	return err
 }
+
+func (conn PostgresConn) HistoricalData(hostname string) (broadcast.HistoricalData, error) {
+	data := broadcast.HistoricalData{}
+	samples, err := conn.db.Query(`SELECT
+		s.Gpu,
+		s.Received,
+		s.MemoryUtilisation,
+		s.GpuUtilisation,
+		s.MemoryUsed,
+		s.FanSpeed,
+		s.Temp,
+		s.MemoryTemp,
+		s.GraphicsVoltage,
+		s.PowerDraw,
+		s.GraphicsClock,
+		s.MaxGraphicsClock,
+		s.MemoryClock,
+		s.MaxMemoryClock,
+		s.InUse,
+		s.UserName
+		FROM Stats s
+		INNER JOIN GPUs g ON g.Uuid = s.Gpu
+		WHERE g.Machine=$1
+		ORDER BY s.Received`,
+		hostname,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// XXX: I hate, HATE what is below this comment, but it has to be done because we
+	//  	need to pass each gpu in its own list to the front-end
+	// NOTE: what we do here is that we accumulate over the stats and package all samples by their insert time
+
+	currtimestamp := time.Now()
+	currpoint := broadcast.HistoricalDataPoint{}
+
+	for samples.Next() {
+		var sample broadcast.GPU
+		var timestamp time.Time
+
+		err = samples.Scan(
+			&sample.Uuid,
+			&timestamp,
+			&sample.MemoryUtilisation,
+			&sample.GPUUtilisation,
+			&sample.MemoryUsed,
+			&sample.FanSpeed,
+			&sample.Temp,
+			&sample.MemoryTemp,
+			&sample.GraphicsVoltage,
+			&sample.PowerDraw,
+			&sample.GraphicsClock,
+			&sample.MaxGraphicsClock,
+			&sample.MemoryClock,
+			&sample.MaxMemoryClock,
+			&sample.InUse,
+			&sample.User,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if timestamp != currtimestamp {
+			// dump current bucket into `data`
+			data = append(data, currpoint)
+			currtimestamp = timestamp
+			currpoint.Timestamp = timestamp.Unix()
+			currpoint.Samples = []broadcast.GPU{}
+		}
+		currpoint.Samples = append(currpoint.Samples, sample)
+	}
+	data = append(data, currpoint)
+
+	return data[1:], nil
+}
+func (conn PostgresConn) AggregateData(days int) (broadcast.AggregateData, error) {
+	// TODO: add functionality for this to be variable
+	var data broadcast.AggregateData
+	threshold := time.Now().AddDate(0, 0, -days).Unix()
+
+	samples, err := conn.db.Query(`SELECT s.Received,
+		s.PowerDraw,
+		s.InUse
+		FROM Stats s
+		WHERE s.Received > $1
+		ORDER BY s.Received`,
+		threshold,
+	)
+	if err != nil {
+		return data, err
+	}
+
+	// XXX: I hate EVERYTHING about the rest of this function,
+	//  	but idk how to write a better SQL query
```

Comment:
I have an idea on how to maybe get this done in the db, can be done in another pr